pkg/controller/controller_test.go

```diff
@@ -705,6 +708,147 @@ func TestController(t *testing.T) {
 				assert.Equal(t, "specification of the created/updated object does not match the desired spec", resCond.Message)
 			},
 		},
+		"binding secret references are resolved": &testCase{
```

Comment:
At least 50% of the time on this PR was trying to figure out this one test... even though I've _already_ looked at them. Might be time to simplify them?

Reply:
Any suggestions?
For readability I think I at least would like to split them into separate files - put each testcase into a separate one into a function that returns a `*testCase`. Then this top level test will just have a map/slice of function invocations to get the test cases.
To avoid cluttering the `controller` directory we can have a sibling `controller_test` package with these tests. It is known as an external test package - it's a special things `go test` knows how to handle. It is a very underdocumented feature. Cannot find a good explanation e.g. it is mentioned here https://dave.cheney.net/2014/12/01/five-suggestions-for-setting-up-a-go-project

> For testing, your test files, those ending with _test.go, may declare themselves to be in the same package, but with _test appended to the package declaration. This is known as an external test. For now, just accept that you can’t put the code for multiple packages into one directory.

Reply:
I'm thinking about it, not to much effect.

One thing that I would like to try - although this is not really addressing the complexity issue here - is to structure tests with helpers rather than iterating over maps. So rather than having 'testName: somerandomparams/functions' in a map, it would be:

```
func testName {
  doBlahTest(t, &BlahTestConfig{
    params as before...
  })
  // now do other asserts (i.e. we can more easily drop some of the 'parameters' from above)
}
```

This means you keep the top level test structure (easy testing of a particular test from IDE @nilebox!), and can reasonably add additional asserts without the abstraction of 'some other thing is processing this config and doing something opaque with it'.

Reply:
Cool, that is how our integration tests are structured. That would be much better 👍